src/borrow_tracker/tree_borrows/foreign_access_skipping.rs

```diff
@@ -0,0 +1,95 @@
+use super::AccessKind;
+use super::tree::AccessRelatedness;
+
+/// To speed up tree traversals, we want to skip traversing subtrees when we know the traversal will have no effect.
+/// This is often the case for foreign accesses, since usually foreign accesses happen several times in a row, but also
+/// foreign accesses are idempotent. In particular, see tests `foreign_read_is_noop_after_foreign_write` and `all_transitions_idempotent`.
+/// Thus, for each node we keep track of the "strongest idempotent foreign access," (SIFA) i.e. which foreign access can be skipped.
+/// This enum represents the kinds of values we store:
+/// `LocalAccess` means that the last access was local, and thus the next foreign access needs to happen, it can not be skipped.
+/// `ForeignRead` means that we are idempotent under foreign reads, but not (yet) under foreign writes.
+/// `ForeignWrite` means that we are idempotent under foreign writes. In other words, this node can be skipped for all foreign accesses.
+/// Due to the subtree property of traversals, this also means that all our children can be skipped, which makes this optimization
+/// worthwhile. In order for this to work, the invariant for multiple nodes in a tree is that at each location, the SIFA at each
+/// child must be stronger than that at the parent. However, note that this is quite invariant, due to retags inserting nodes across
+/// the entire range, which can violate this invariant without explicitly causing a local access which would reset things.
+/// So this needs to be done manually, thus `ensure_no_stronger_than`, and also the test `permission_sifa_is_correct` in `perms.rs`
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Default)]
```

Comment:
Seems worth mentioning that since we derive `PartialOrd`, the order in which we write the elements here is crucial! They must be order from weakest to strongest.

Reply:
there's a test for that.

Reply:
Tests don't replace comments.